src/Hamiltonians/GutzwillerSampling.jl

```diff
@@ -0,0 +1,102 @@
+"""
+    GutzwillerSampling(::AbstractHamiltonian; g)
+
+Wrapper over any `AbstractHamiltonian` that implements Gutzwiller sampling. In this
+importance sampling scheme the Hamiltonian is modified as follows.
+
+```math
+\\tilde{H}_{ij} = H_{ij} e^{-g(H_{jj} - H_{ii})}
+```
+
+# Example
+
+```jldoctest
+julia> H = HubbardMom1D(BoseFS{3}((1,1,1)); u=6.0, t=1.0)
+HubbardMom1D(BoseFS{3,3}((1, 1, 1)); u=6.0, t=1.0)
+
+julia> I = GutzwillerSampling(H, g=0.3)
+GutzwillerSampling(HubbardMom1D(BoseFS{3,3}((1, 1, 1)); u=6.0, t=1.0); g=0.3)
+
+julia> get_offdiagonal(H, BoseFS((2, 1, 0)), 1)
+(BoseFS{3,3}((1, 0, 2)), 2.0)
+
+julia> get_offdiagonal(I, BoseFS((2, 1, 0)), 1)
+(BoseFS{3,3}((1, 0, 2)), 0.8131393194811987)
+```
+"""
+struct GutzwillerSampling{A,T,H<:AbstractHamiltonian{T},G} <: AbstractHamiltonian{T}
+    hamiltonian::H
+end
+
+function GutzwillerSampling(h, g)
+    G = eltype(h)(g)
+    return GutzwillerSampling{false,eltype(h),typeof(h),G}(h)
+end
+GutzwillerSampling(h; g) = GutzwillerSampling(h, g)
+
+function Base.show(io::IO, h::GutzwillerSampling{A}) where {A}
+    A && print(io, "adjoint(")
+    print(io, "GutzwillerSampling(", h.hamiltonian, "; g=", h.g, ")")
+    A && print(io, ")")
+end
+
+starting_address(h::GutzwillerSampling) = starting_address(h.hamiltonian)
+
+LOStructure(::Type{<:GutzwillerSampling{<:Any,<:Any,H}}) where {H} = _lo_str(LOStructure(H))
+_lo_str(::LOStructure) = AdjointKnown()
+_lo_str(::AdjointUnknown) = AdjointUnknown()
+
+function LinearAlgebra.adjoint(h::GutzwillerSampling{A,T,<:Any,G}) where {A,T,G}
+    h_adj = h.hamiltonian'
+    return GutzwillerSampling{!A,T,typeof(h_adj),G}(h_adj)
+end
+
+Base.getproperty(h::GutzwillerSampling, s::Symbol) = getproperty(h, Val(s))
+Base.getproperty(h::GutzwillerSampling{<:Any,<:Any,<:Any,G}, ::Val{:g}) where G = G
+Base.getproperty(h::GutzwillerSampling, ::Val{:hamiltonian}) = getfield(h, :hamiltonian)
```

Comment:
Hmm, maybe information on how to use this functionality should appear in docstrings, though. The best place would probably be the docstring of `GutzwillerSampling`.

Reply:
Added to docstring. Ditto for `GuidingVectorSampling`.